src/main.rs

```diff
@@ -0,0 +1,188 @@
+
+    let output = match get_gl_info(template, display_builder, &event_loop) {
+        Ok(gl_info) => format!("{}", gl_info),
+        Err(err) => format!("ERROR: {}", err),
+    };
+
+    println!("{}", output);
+
+    if args.len() == 3 && args[1] == "-f" {
+        let filename = &args[2];
+        std::fs::write(filename, output.as_bytes())?;
+    }
+
+    Ok(())
+}
+
+fn get_gl_info(
+    template: ConfigTemplateBuilder,
+    display_builder: DisplayBuilder,
+    event_loop: &EventLoop<()>,
+) -> Result<GLInfo, String> {
+    // We just created the event loop, so initialize the display, pick the config, and
+    // create the context.
+    let (window, gl_config) =
+        match display_builder
+            .clone()
+            .build(event_loop, template.clone(), |mut configs| {
+                configs.next().expect("Could not get any configs")
+            }) {
+            Ok((window, gl_config)) => (window.unwrap(), gl_config),
+            Err(err) => {
+                return Err(err.to_string());
+            }
+        };
+
+    // Create gl context.
+    let (driver, context) = create_gl_context(&window, &gl_config)?;
+    let gl_context = Some(context.treat_as_possibly_current());
+
+    let attrs = window
+        .build_surface_attributes(Default::default())
+        .map_err(|err| format!("Failed to build surface attributes: {err}"))?;
+    let gl_surface = unsafe {
+        gl_config
+            .display()
+            .create_window_surface(&gl_config, &attrs)
+            .map_err(|err| format!("Failed to create a window surface: {err}"))?
+    };
+
+    // The context needs to be current for the Renderer to set up shaders and
+    // buffers. It also performs function loading, which needs a current context on
+    // WGL.
+    let gl_context = gl_context.as_ref().ok_or("Failed to get a GL context")?;
+    gl_context.make_current(&gl_surface).unwrap();
+
+    let gl_display = gl_config.display();
+
+    let gl = gl::Gl::load_with(|symbol| {
+        let symbol = CString::new(symbol).unwrap();
+        gl_display.get_proc_address(symbol.as_c_str()).cast()
+    });
+
+    let mut gl_info = GLInfo {
+        driver: driver.to_string(),
+        vendor: "".to_string(),
+        renderer: "".to_string(),
+        version: "".to_string(),
+        shading_language: "".to_string(),
+    };
+    if let Some(vendor) = get_gl_string(&gl, gl::VENDOR) {
+        gl_info.vendor = vendor.to_string_lossy().into();
+    }
+    if let Some(renderer) = get_gl_string(&gl, gl::RENDERER) {
+        gl_info.renderer = renderer.to_string_lossy().into();
+    }
+    if let Some(version) = get_gl_string(&gl, gl::VERSION) {
+        gl_info.version = version.to_string_lossy().into();
+    }
+
+    if let Some(shaders_version) = get_gl_string(&gl, gl::SHADING_LANGUAGE_VERSION) {
+        gl_info.shading_language = shaders_version.to_string_lossy().into();
+    }
+    Ok(gl_info)
```

Comment:
You could make this a function `GLInfo::load(&gl::Gl)` (or `from_gl`, `read`...)

Reply:
Great suggestion. Also simplified `get_gl_string`.